warehouse/query-core/src/main/java/datawave/query/tables/chained/iterators/LazyLoadingRangesIterator.java

```diff
@@ -0,0 +1,231 @@
+    private final int numRangesPerScanner;
+
+    final ChunkSizeEncoding chunkSizeEncoder = new ChunkSizeEncoding();
+    IntegerEncoding bucketEncoder;
+    int indexBuckets;
+
+    public LazyLoadingRangesIterator(SSDeepSimilarityQueryConfiguration config, ScannerFactory scannerFactory) throws TableNotFoundException {
+        queryMapKeysIterator = config.getState().getQueryMap().keys().iterator();
+        bucketEncoder = new IntegerEncoding(config.getBucketEncodingBase(), config.getBucketEncodingLength());
+        indexBuckets = config.getIndexBuckets();
+        scanner = scannerFactory.newScanner(config.getTableName(), config.getAuthorizations(), config.getQueryThreads(), config.getQuery());
+        numRangesPerScanner = config.getNumRangesPerScanner();
+    }
+
+    /**
+     * Continue creating new iterators using a new calculated set of ranges until a next result can be found
+     *
+     * @return true if a result has been found, false if not
+     */
+    @Override
+    public boolean hasNext() {
+        while (scannerIterator == null || !scannerIterator.hasNext()) {
+            if (queryMapKeysIterator.hasNext()) {
+                scannerIterator = createIteratorWithNewRanges();
+            } else {
+                return false;
+            }
+        }
+
+        return scannerIterator.hasNext();
+    }
+
+    /**
+     * Continue creating new iterators using a new calculated set of ranges until a next result can be found
+     *
+     * @return the next item
+     */
+    @Override
+    public Map.Entry<Key,Value> next() {
+        while (scannerIterator == null || !scannerIterator.hasNext()) {
+            scannerIterator = createIteratorWithNewRanges();
+        }
+
+        return scannerIterator.next();
+    }
+
+    /**
+     * Process the query map keys to create ranges to scan in accumulo. The number of ranges generated at a time is determined by {@numRangesPerScanner}. The
+     * number of ranges may go over numRangesPerScanner slightly due to processing all index buckets per NGramTuple
+     *
+     * @return an iterator based on the next set of calculated ranges
+     */
+    private Iterator<Map.Entry<Key,Value>> createIteratorWithNewRanges() {
+        final Collection<Range> ranges = new HashSet<>();
+
+        while (queryMapKeysIterator.hasNext() && ranges.size() <= numRangesPerScanner) {
+            NGramTuple ct = queryMapKeysIterator.next();
+            final String sizeAndChunk = chunkSizeEncoder.encode(ct.getChunkSize()) + ct.getChunk();
+            for (int i = 0; i < indexBuckets; i++) {
+                final String bucketedSizeAndChunk = bucketEncoder.encode(i) + sizeAndChunk;
+                ranges.add(Range.exact(new Text(bucketedSizeAndChunk)));
+            }
+        }
+
+        scanner.setRanges(ranges);
+
+        log.info("Lazy loaded " + ranges.size() + " ranges.");
+        if (log.isDebugEnabled()) {
+            log.debug("Ranges are: " + ranges);
+        }
```

Comment:
Consider using slf4j instead? Consider making the statements debug and trace level?
```suggestion
        log.debug("Lazy loaded {} ranges.", ranges.size());
        log.trace("Ranges are: {}",ranges);        
```

Reply:
I switched to using slf4j and made the log level changes.